File: NFComm/NFCore/NFCObject.cpp
// -------------------------------------------------------------------------
//    @FileName         :    NFCObject.cpp
//    @Author           :    LvSheng.Huang
//    @Date             :    2012-03-01
//    @Module           :    NFCObject
//
// -------------------------------------------------------------------------

#include "NFCObject.h"
#include "NFCRecordManager.h"
#include "NFCPropertyManager.h"

NFCObject::NFCObject(NFGUID self, NFIPluginManager* pLuginManager)
    : NFIObject(self)
{
	mObjectEventState = COE_CREATE_NODATA;

    mSelf = self;
    m_pPluginManager = pLuginManager;

    m_pRecordManager = NF_SHARE_PTR<NFCRecordManager>(NF_NEW NFCRecordManager(mSelf));
    m_pPropertyManager = NF_SHARE_PTR<NFCPropertyManager>(NF_NEW NFCPropertyManager(mSelf));
}

NFCObject::~NFCObject()
{

}

bool NFCObject::Init()
{
    return true;
}

bool NFCObject::Shut()
{
    return true;
}

bool NFCObject::Execute()
{
    return true;
}

bool NFCObject::AddRecordCallBack(const std::string& strRecordName, const RECORD_EVENT_FUNCTOR_PTR& cb)
{
    NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
    if (pRecord)
    {
        pRecord->AddRecordHook(cb);

        return true;
    }

    return false;
}

bool NFCObject::AddPropertyCallBack(const std::string& strCriticalName, const PROPERTY_EVENT_FUNCTOR_PTR& cb)
{
    NF_SHARE_PTR<NFIProperty> pProperty = GetPropertyManager()->GetElement(strCriticalName);
    if (pProperty)
    {
        pProperty->RegisterCallback(cb);

        return true;
    }

    return false;
}

CLASS_OBJECT_EVENT NFCObject::GetState()
{
	return mObjectEventState;
}

bool NFCObject::SetState(const CLASS_OBJECT_EVENT eState)
{
	mObjectEventState = eState;
	return true;
}

bool NFCObject::FindProperty(const std::string& strPropertyName)
{
    if (GetPropertyManager()->GetElement(strPropertyName))
    {
        return true;
    }

    return false;
}

bool NFCObject::SetPropertyInt(const std::string& strPropertyName, const NFINT64 nValue)
{
    NF_SHARE_PTR<NFIProperty> pProperty = GetPropertyManager()->GetElement(strPropertyName);
    if (pProperty)
    {
        return pProperty->SetInt(nValue);
    }

    return false;
}

bool NFCObject::SetPropertyFloat(const std::string& strPropertyName, const double dwValue)
{
    NF_SHARE_PTR<NFIProperty> pProperty = GetPropertyManager()->GetElement(strPropertyName);
    if (pProperty)
    {
        return pProperty->SetFloat(dwValue);
    }

    return false;
}

bool NFCObject::SetPropertyString(const std::string& strPropertyName, const std::string& strValue)
{
    NF_SHARE_PTR<NFIProperty> pProperty = GetPropertyManager()->GetElement(strPropertyName);
    if (pProperty)
    {
        return pProperty->SetString(strValue);
    }

    return false;
}

bool NFCObject::SetPropertyObject(const std::string& strPropertyName, const NFGUID& obj)
{
    NF_SHARE_PTR<NFIProperty> pProperty = GetPropertyManager()->GetElement(strPropertyName);
    if (pProperty)
    {
        return pProperty->SetObject(obj);
    }

    return false;
}

bool NFCObject::SetPropertyVector2(const std::string& strPropertyName, const NFVector2& value)
{
	NF_SHARE_PTR<NFIProperty> pProperty = GetPropertyManager()->GetElement(strPropertyName);
	if (pProperty)
	{
		return pProperty->SetVector2(value);
	}

	return false;
}

bool NFCObject::SetPropertyVector3(const std::string& strPropertyName, const NFVector3& value)
{
	NF_SHARE_PTR<NFIProperty> pProperty = GetPropertyManager()->GetElement(strPropertyName);
	if (pProperty)
	{
		return pProperty->SetVector3(value);
	}

	return false;
}

NFINT64 NFCObject::GetPropertyInt(const std::string& strPropertyName)
{
    NF_SHARE_PTR<NFIProperty> pProperty = GetPropertyManager()->GetElement(strPropertyName);
    if (pProperty)
    {
        return pProperty->GetInt();
    }

    return 0;
}

int NFCObject::GetPropertyInt32(const std::string& strPropertyName)
{
	NF_SHARE_PTR<NFIProperty> pProperty = GetPropertyManager()->GetElement(strPropertyName);
	if (pProperty)
	{
		return pProperty->GetInt32();
	}

	return 0;
}

double NFCObject::GetPropertyFloat(const std::string& strPropertyName)
{
    NF_SHARE_PTR<NFIProperty> pProperty = GetPropertyManager()->GetElement(strPropertyName);
    if (pProperty)
    {
        return pProperty->GetFloat();
    }

    return 0.0;
}

const std::string& NFCObject::GetPropertyString(const std::string& strPropertyName)
{
    NF_SHARE_PTR<NFIProperty> pProperty = GetPropertyManager()->GetElement(strPropertyName);
    if (pProperty)
    {
        return pProperty->GetString();
    }

    return NULL_STR;
}

const NFGUID& NFCObject::GetPropertyObject(const std::string& strPropertyName)
{
    NF_SHARE_PTR<NFIProperty> pProperty = GetPropertyManager()->GetElement(strPropertyName);
    if (pProperty)
    {
        return pProperty->GetObject();
    }

    return NULL_OBJECT;
}

const NFVector2& NFCObject::GetPropertyVector2(const std::string& strPropertyName)
{
	NF_SHARE_PTR<NFIProperty> pProperty = GetPropertyManager()->GetElement(strPropertyName);
	if (pProperty)
	{
		return pProperty->GetVector2();
	}

	return NULL_VECTOR2;
}

const NFVector3& NFCObject::GetPropertyVector3(const std::string& strPropertyName)
{
	NF_SHARE_PTR<NFIProperty> pProperty = GetPropertyManager()->GetElement(strPropertyName);
	if (pProperty)
	{
		return pProperty->GetVector3();
	}

	return NULL_VECTOR3;
}

bool NFCObject::FindRecord(const std::string& strRecordName)
{
    NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
    if (pRecord)
    {
        return true;
    }

    return false;
}

bool NFCObject::SetRecordInt(const std::string& strRecordName, const int nRow, const int nCol, const NFINT64 nValue)
{
    NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
    if (pRecord)
    {
        return pRecord->SetInt(nRow, nCol, nValue);
    }

    return false;
}

bool NFCObject::SetRecordInt(const std::string& strRecordName, const int nRow, const std::string& strColTag, const NFINT64 value)
{
    NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
    if (pRecord)
    {
        return pRecord->SetInt(nRow, strColTag, value);
    }

    return false;
}

bool NFCObject::SetRecordFloat(const std::string& strRecordName, const int nRow, const int nCol, const double dwValue)
{
    NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
    if (pRecord)
    {
        return pRecord->SetFloat(nRow, nCol, dwValue);
    }

    return false;
}

bool NFCObject::SetRecordFloat(const std::string& strRecordName, const int nRow, const std::string& strColTag, const double value)
{
    NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
    if (pRecord)
    {
        return pRecord->SetFloat(nRow, strColTag, value);
    }

    return false;
}

bool NFCObject::SetRecordString(const std::string& strRecordName, const int nRow, const int nCol, const std::string& strValue)
{
    NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
    if (pRecord)
    {
        return pRecord->SetString(nRow, nCol, strValue.c_str());
    }

    return false;
}

bool NFCObject::SetRecordString(const std::string& strRecordName, const int nRow, const std::string& strColTag, const std::string& value)
{
    NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
    if (pRecord)
    {
        return pRecord->SetString(nRow, strColTag, value.c_str());
    }

    return false;
}

bool NFCObject::SetRecordObject(const std::string& strRecordName, const int nRow, const int nCol, const NFGUID& obj)
{
    NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
    if (pRecord)
    {
        return pRecord->SetObject(nRow, nCol, obj);
    }

    return false;
}

bool NFCObject::SetRecordObject(const std::string& strRecordName, const int nRow, const std::string& strColTag, const NFGUID& value)
{
    NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
    if (pRecord)
    {
        return pRecord->SetObject(nRow, strColTag, value);
    }

    return false;
}

bool NFCObject::SetRecordVector2(const std::string& strRecordName, const int nRow, const int nCol, const NFVector2& obj)
{
	NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
	if (pRecord)
	{
		return pRecord->SetVector2(nRow, nCol, obj);
	}

	return false;
}

bool NFCObject::SetRecordVector2(const std::string& strRecordName, const int nRow, const std::string& strColTag, const NFVector2& value)
{
	NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
	if (pRecord)
	{
		return pRecord->SetVector2(nRow, strColTag, value);
	}

	return false;
}

bool NFCObject::SetRecordVector3(const std::string& strRecordName, const int nRow, const int nCol, const NFVector3& obj)
{
	NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
	if (pRecord)
	{
		return pRecord->SetVector3(nRow, nCol, obj);
	}

	return false;
}

bool NFCObject::SetRecordVector3(const std::string& strRecordName, const int nRow, const std::string& strColTag, const NFVector3& value)
{
	NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
	if (pRecord)
	{
		return pRecord->SetVector3(nRow, strColTag, value);
	}

	return false;
}

NFINT64 NFCObject::GetRecordInt(const std::string& strRecordName, const int nRow, const int nCol)
{
    NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
    if (pRecord)
    {
        return pRecord->GetInt(nRow, nCol);
    }

    return 0;
}

NFINT64 NFCObject::GetRecordInt(const std::string& strRecordName, const int nRow, const std::string& strColTag)
{
    NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
    if (pRecord)
    {
        return pRecord->GetInt(nRow, strColTag);
    }

    return 0;
}

double NFCObject::GetRecordFloat(const std::string& strRecordName, const int nRow, const int nCol)
{
    NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
    if (pRecord)
    {
        return pRecord->GetFloat(nRow, nCol);
    }

    return 0.0;
}

double NFCObject::GetRecordFloat(const std::string& strRecordName, const int nRow, const std::string& strColTag)
{
    NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
    if (pRecord)
    {
        return pRecord->GetFloat(nRow, strColTag);
    }

    return 0.0;
}

const std::string& NFCObject::GetRecordString(const std::string& strRecordName, const int nRow, const int nCol)
{
    NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
    if (pRecord)
    {
        return pRecord->GetString(nRow, nCol);
    }

    return NULL_STR;
}

const std::string& NFCObject::GetRecordString(const std::string& strRecordName, const int nRow, const std::string& strColTag)
{
    NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
    if (pRecord)
    {
        return pRecord->GetString(nRow, strColTag);
    }

    return NULL_STR;
}

const NFGUID& NFCObject::GetRecordObject(const std::string& strRecordName, const int nRow, const int nCol)
{
    NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
    if (pRecord)
    {
        return pRecord->GetObject(nRow, nCol);
    }

    return NULL_OBJECT;
}

const NFGUID& NFCObject::GetRecordObject(const std::string& strRecordName, const int nRow, const std::string& strColTag)
{
    NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
    if (pRecord)
    {
        return pRecord->GetObject(nRow, strColTag);
    }

    return NULL_OBJECT;
}

const NFVector2& NFCObject::GetRecordVector2(const std::string& strRecordName, const int nRow, const int nCol)
{
	NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
	if (pRecord)
	{
		return pRecord->GetVector2(nRow, nCol);
	}

	return NULL_VECTOR2;
}

const NFVector2& NFCObject::GetRecordVector2(const std::string& strRecordName, const int nRow, const std::string& strColTag)
{
	NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
	if (pRecord)
	{
		return pRecord->GetVector2(nRow, strColTag);
	}

	return NULL_VECTOR2;
}

const NFVector3& NFCObject::GetRecordVector3(const std::string& strRecordName, const int nRow, const int nCol)
{
	NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
	if (pRecord)
	{
		return pRecord->GetVector3(nRow, nCol);
	}

	return NULL_VECTOR3;
}

const NFVector3& NFCObject::GetRecordVector3(const std::string& strRecordName, const int nRow, const std::string& strColTag)
{
	NF_SHARE_PTR<NFIRecord> pRecord = GetRecordManager()->GetElement(strRecordName);
	if (pRecord)
	{
		return pRecord->GetVector3(nRow, strColTag);
	}

	return NULL_VECTOR3;
}

NF_SHARE_PTR<NFIRecordManager> NFCObject::GetRecordManager()
{
    return m_pRecordManager;
}

NF_SHARE_PTR<NFIPropertyManager> NFCObject::GetPropertyManager()
{
    return m_pPropertyManager;
}

NFGUID NFCObject::Self()
{
    return mSelf;
}

File: NFClient/Unity3D/Assets/_NF/NFrame/Object/NFCObject.cs
//-----------------------------------------------------------------------
// <copyright file="NFCObject.cs">
//     Copyright (C) 2015-2015 lvsheng.huang <https://github.com/ketoo/NFrame>
// </copyright>
//-----------------------------------------------------------------------
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace NFrame
{
    public class NFCObject : NFIObject
    {
		public NFCObject(NFGUID self, int nContainerID, int nGroupID, string strClassName, string strConfigIndex)
		{
			mSelf = self;
            mstrClassName = strClassName;
            mstrConfigIndex = strConfigIndex;
			mnContainerID = nContainerID;
			mnGroupID = nGroupID;
			Init();
		}

		~NFCObject()
		{
			Shut();
		}

        public override void Init()
        {
			mRecordManager = new NFCRecordManager(mSelf);
			mPropertyManager = new NFCPropertyManager(mSelf);
			mHeartManager = new NFCHeartBeatManager(mSelf);

            return;
        }

        public override void Shut()
        {
            NFIDataList xRecordList = mRecordManager.GetRecordList();
            if (null != xRecordList)
            {
                for(int i = 0; i < xRecordList.Count(); ++i)
                {
                    string strRecordName = xRecordList.StringVal(i);
                    NFIRecord xRecord = mRecordManager.GetRecord(strRecordName);
                    if (null !=  xRecord)
                    {
                        xRecord.Clear();
                    }
                }
            }

			mRecordManager = null;
			mPropertyManager = null;
            mHeartManager = null;

            return;
        }

        public override void AddHeartBeat(string strHeartBeatName, float fTime, int nCount, NFIHeartBeat.HeartBeatEventHandler handler)
        {
            GetHeartBeatManager().AddHeartBeat(strHeartBeatName, fTime, nCount, handler);
        }

        public override bool FindHeartBeat(string strHeartBeatName)
        {
            return GetHeartBeatManager().FindHeartBeat(strHeartBeatName);
        }

        public override void RemoveHeartBeat(string strHeartBeatName)
        {
            GetHeartBeatManager().RemoveHeartBeat(strHeartBeatName);
        }

        public override bool UpData(float fLastTime, float fAllTime)
        {
            GetHeartBeatManager().Update(fLastTime);

            return true;
        }

        ///////////////////////////////////////////////////////////////////////
        public override NFGUID Self()
        {
			return mSelf;
        }
		
		public override int ContainerID()
        {
			return mnContainerID;
        }
		
		public override int GroupID()
        {
			return mnGroupID;
        }
		
        public override string ClassName()
        {
            return mstrClassName;
        }

        public override string ConfigIndex()
        {
            return mstrConfigIndex;
        }

        public override bool FindProperty(string strPropertyName)
        {
			if (null != mPropertyManager.GetProperty(strPropertyName))
			{
				return true;
			}

            return false;
        }

        public override bool SetPropertyInt(string strPropertyName, Int64 nValue)
        {
			NFIProperty property = mPropertyManager.GetProperty(strPropertyName);
			if (null == property)
			{
                NFIDataList valueList = new NFCDataList();
                valueList.AddInt(0);
                property = mPropertyManager.AddProperty(strPropertyName, valueList);
            }

			property.SetInt(nValue);
			return true;
        }

        public override bool SetPropertyFloat(string strPropertyName, double fValue)
        {
			NFIProperty property = mPropertyManager.GetProperty(strPropertyName);
			if (null == property)
			{
                NFIDataList valueList = new NFCDataList();
                valueList.AddFloat(0.0f);
                property = mPropertyManager.AddProperty(strPropertyName, valueList);
            }
				
            property.SetFloat(fValue);
			return true;
        }

        public override bool SetPropertyString(string strPropertyName, string strValue)
        {
			NFIProperty property = mPropertyManager.GetProperty(strPropertyName);
			if (null == property)
			{
                NFIDataList valueList = new NFCDataList();
                valueList.AddString("");
                property = mPropertyManager.AddProperty(strPropertyName, valueList);
            }

			property.SetString(strValue);
			return true;
        }

        public override bool SetPropertyObject(string strPropertyName, NFGUID obj)
        {
			NFIProperty property = mPropertyManager.GetProperty(strPropertyName);
			if (null == property)
			{
                NFIDataList valueList = new NFCDataList();
                valueList.AddObject(new NFGUID());
                property = mPropertyManager.AddProperty(strPropertyName, valueList);
            }

			property.SetObject(obj);
			return true;

        }

        public override bool SetPropertyVector2(string strPropertyName, NFVector2 obj)
        {
            NFIProperty property = mPropertyManager.GetProperty(strPropertyName);
            if (null == property)
            {
                NFIDataList valueList = new NFCDataList();
                valueList.AddVector2(new NFVector2());
                property = mPropertyManager.AddProperty(strPropertyName, valueList);
            }

            property.SetVector2(obj);
            return true;

        }

        public override bool SetPropertyVector3(string strPropertyName, NFVector3 obj)
        {
            NFIProperty property = mPropertyManager.GetProperty(strPropertyName);
            if (null == property)
            {
                NFIDataList valueList = new NFCDataList();
                valueList.AddVector3(new NFVector3());
                property = mPropertyManager.AddProperty(strPropertyName, valueList);
            }

            property.SetVector3(obj);
            return true;

        }

        public override Int64 QueryPropertyInt(string strPropertyName)
        {
			NFIProperty property = mPropertyManager.GetProperty(strPropertyName);
			if (null != property)
			{
				return property.QueryInt();
			}

            return 0;
        }

        public override double QueryPropertyFloat(string strPropertyName)
        {
			NFIProperty property = mPropertyManager.GetProperty(strPropertyName);
			if (null != property)
			{
				return property.QueryFloat();
			}

            return 0.0;
        }

        public override string QueryPropertyString(string strPropertyName)
        {
			NFIProperty property = mPropertyManager.GetProperty(strPropertyName);
			if (null != property)
			{
				return property.QueryString();
			}

            return NFIDataList.NULL_STRING;
        }

        public override NFGUID QueryPropertyObject(string strPropertyName)
        {
			NFIProperty property = mPropertyManager.GetProperty(strPropertyName);
			if (null != property)
			{
				return property.QueryObject();
			}

            return NFIDataList.NULL_OBJECT;
        }

        public override NFVector2 QueryPropertyVector2(string strPropertyName)
        {
            NFIProperty property = mPropertyManager.GetProperty(strPropertyName);
            if (null != property)
            {
                return property.QueryVector2();
            }

            return NFIDataList.NULL_VECTOR2;
        }

        public override NFVector3 QueryPropertyVector3(string strPropertyName)
        {
            NFIProperty property = mPropertyManager.GetProperty(strPropertyName);
            if (null != property)
            {
                return property.QueryVector3();
            }

            return NFIDataList.NULL_VECTOR3;
        }

        public override bool FindRecord(string strRecordName)
        {
			NFIRecord record = mRecordManager.GetRecord(strRecordName);
			if (null != record)
			{
				return true;
			}
            return false;
        }

        public override bool SetRecordInt(string strRecordName, int nRow, int nCol, Int64 nValue)
        {
			NFIRecord record = mRecordManager.GetRecord(strRecordName);
			if (null != record)
			{
				record.SetInt(nRow, nCol, nValue);
				return true;
			}

			return false;
        }

        public override bool SetRecordFloat(string strRecordName, int nRow, int nCol, double fValue)
        {
			NFIRecord record = mRecordManager.GetRecord(strRecordName);
			if (null != record)
			{
				record.SetFloat(nRow, nCol, fValue);
				return true;
			}

			return false;
        }

        public override bool SetRecordString(string strRecordName, int nRow, int nCol, string strValue)
        {
			NFIRecord record = mRecordManager.GetRecord(strRecordName);
			if (null != record)
			{
				record.SetString(nRow, nCol, strValue);
				return true;
			}

			return false;
        }

        public override bool SetRecordObject(string strRecordName, int nRow, int nCol, NFGUID obj)
        {
			NFIRecord record = mRecordManager.GetRecord(strRecordName);
			if (null != record)
			{
				record.SetObject(nRow, nCol, obj);
				return true;
			}

			return false;
        }

        public override bool SetRecordVector2(string strRecordName, int nRow, int nCol, NFVector2 obj)
        {
            NFIRecord record = mRecordManager.GetRecord(strRecordName);
            if (null != record)
            {
                record.SetVector2(nRow, nCol, obj);
                return true;
            }

            return false;
        }

        public override bool SetRecordVector3(string strRecordName, int nRow, int nCol, NFVector3 obj)
        {
            NFIRecord record = mRecordManager.GetRecord(strRecordName);
            if (null != record)
            {
                record.SetVector3(nRow, nCol, obj);
                return true;
            }

            return false;
        }

        public override Int64 QueryRecordInt(string strRecordName, int nRow, int nCol)
        {
			NFIRecord record = mRecordManager.GetRecord(strRecordName);
			if (null != record)
			{
				return record.QueryInt(nRow, nCol);
			}

            return 0;
        }

        public override double QueryRecordFloat(string strRecordName, int nRow, int nCol)
        {
			NFIRecord record = mRecordManager.GetRecord(strRecordName);
			if (null != record)
			{
				return record.QueryFloat(nRow, nCol);
			}

            return 0.0;
        }

        public override string QueryRecordString(string strRecordName, int nRow, int nCol)
        {
			NFIRecord record = mRecordManager.GetRecord(strRecordName);
			if (null != record)
			{
				return record.QueryString(nRow, nCol);
			}

            return NFIDataList.NULL_STRING;
        }

        public override NFGUID QueryRecordObject(string strRecordName, int nRow, int nCol)
        {
			NFIRecord record = mRecordManager.GetRecord(strRecordName);
			if (null != record)
			{
				return record.QueryObject(nRow, nCol);
			}

            return null;
        }

        public override NFVector2 QueryRecordVector2(string strRecordName, int nRow, int nCol)
        {
            NFIRecord record = mRecordManager.GetRecord(strRecordName);
            if (null != record)
            {
                return record.QueryVector2(nRow, nCol);
            }

            return null;
        }

        public override NFVector3 QueryRecordVector3(string strRecordName, int nRow, int nCol)
        {
            NFIRecord record = mRecordManager.GetRecord(strRecordName);
            if (null != record)
            {
                return record.QueryVector3(nRow, nCol);
            }

            return null;
        }

        public override NFIRecordManager GetRecordManager()
        {
			return mRecordManager;
        }

        public override NFIHeartBeatManager GetHeartBeatManager()
        {
			return mHeartManager;
        }

        public override NFIPropertyManager GetPropertyManager()
        {
			return mPropertyManager;
        }

		NFGUID mSelf;
		int mnContainerID;
		int mnGroupID;
		
		string mstrClassName;
		string mstrConfigIndex;

		NFIRecordManager mRecordManager;
		NFIPropertyManager mPropertyManager;
		NFIHeartBeatManager mHeartManager;
	}
}